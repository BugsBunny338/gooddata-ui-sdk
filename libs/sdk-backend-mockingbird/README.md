# Mocking bird

This package contains test support and mock implementations of various concepts in the sdk-backend-spi.
