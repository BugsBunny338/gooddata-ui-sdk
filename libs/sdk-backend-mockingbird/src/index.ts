// (C) 2019-2020 GoodData Corporation

import {
    dummyBackend,
    dummyBackendEmptyData,
    dummyDataView,
    dummyDataFacade,
} from "@gooddata/sdk-backend-base";
export { dummyBackend, dummyBackendEmptyData, dummyDataView, dummyDataFacade };

export { recordedBackend } from "./recordedBackend";

export {
    RecordingIndex,
    ExecutionRecording,
    InsightRecording,
    DisplayFormRecording,
    ScenarioRecording,
    RecordedBackendConfig,
} from "./recordedBackend/types";

export {
    recordedDataView,
    DataViewAll,
    dataViewWindow,
    DataViewFirstPage,
} from "./recordedBackend/execution";

export {
    LegacyExecutionRecording,
    legacyRecordedBackend,
    legacyRecordedDataFacade,
} from "./legacyRecordedBackend";
