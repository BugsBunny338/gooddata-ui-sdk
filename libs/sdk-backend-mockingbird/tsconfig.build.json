{
    "extends": "./tsconfig.json",
    "compilerOptions": {
        "baseUrl": null,
        "paths": null
    },
    "include": ["src/index.ts"]
}
