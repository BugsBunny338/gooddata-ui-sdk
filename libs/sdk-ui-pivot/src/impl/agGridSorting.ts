// (C) 2007-2020 GoodData Corporation
import {
    getAttributeLocators,
    getIdsFromUri,
    getLastFieldId,
    getLastFieldType,
    getParsedFields,
} from "./agGridUtils";
import { FIELD_TYPE_ATTRIBUTE, FIELD_TYPE_MEASURE } from "./agGridConst";
import { assortDimensionDescriptors } from "./agGridHeaders";
import { ISortModelItem } from "./agGridTypes";
import {
    IAttributeSortItem,
    IMeasureSortItem,
    isAttributeAreaSort,
    isAttributeSort,
    ISortItem,
    newAttributeAreaSort,
    newAttributeSort,
    SortDirection,
} from "@gooddata/sdk-model";
import { IExecutionResult } from "@gooddata/sdk-backend-spi";
import invariant, { InvariantError } from "ts-invariant";

/*
 * All code related to sorting the ag-grid backed Pivot Table is concentrated here
 */

export const getSortItemByColId = (
    result: IExecutionResult,
    colId: string,
    direction: SortDirection,
    originalSorts: ISortItem[] = [],
): IMeasureSortItem | IAttributeSortItem => {
    const { dimensions } = result;

    const fields = getParsedFields(colId);
    const lastFieldType = getLastFieldType(fields);
    const lastFieldId = getLastFieldId(fields);

    // search columns first when sorting in columns to use the proper header
    // in case the same attribute is in both rows and columns
    const searchDimensionIndex = lastFieldType === FIELD_TYPE_MEASURE ? 1 : 0;
    const { attributeDescriptors, measureDescriptors } = assortDimensionDescriptors([
        dimensions[searchDimensionIndex],
    ]);

    if (lastFieldType === FIELD_TYPE_ATTRIBUTE) {
        for (const header of attributeDescriptors) {
            if (getIdsFromUri(header.attributeHeader.uri)[0] === lastFieldId) {
                const attributeLocalId = header.attributeHeader.localIdentifier;

                // try to find the original sort item in case it had an aggregation set so we can keep it in (RAIL-1992)
                // we intentionally ignore the direction to make sure the UX is predictable
                const matchingOriginalSortItem = originalSorts.find(
                    (s) => isAttributeSort(s) && s.attributeSortItem.attributeIdentifier === attributeLocalId,
                ) as IAttributeSortItem;

                return isAttributeAreaSort(matchingOriginalSortItem)
                    ? newAttributeAreaSort(attributeLocalId, direction)
                    : newAttributeSort(attributeLocalId, direction);
            }
        }
        throw new InvariantError(`could not find attribute header matching ${colId}`);
    } else if (lastFieldType === FIELD_TYPE_MEASURE) {
        const headerItem = measureDescriptors[parseInt(lastFieldId, 10)];
        const attributeLocators = getAttributeLocators(fields, attributeDescriptors);

        return {
            measureSortItem: {
                direction,
                locators: [
                    ...attributeLocators,
                    {
                        measureLocatorItem: {
                            measureIdentifier: headerItem.measureHeaderItem.localIdentifier,
                        },
                    },
                ],
            },
        };
    }
    throw new InvariantError(`could not find header matching ${colId}`);
};

export function getSortsFromModel(sortModel: ISortModelItem[], result: IExecutionResult): ISortItem[] {
    const originalSorts = result.definition.sortBy;

    return sortModel.map((sortModelItem: ISortModelItem) => {
        const { colId, sort } = sortModelItem;
        const sortHeader = getSortItemByColId(result, colId, sort, originalSorts);
        invariant(sortHeader, `unable to find sort item by field ${colId}`);
        return sortHeader;
    });
}
