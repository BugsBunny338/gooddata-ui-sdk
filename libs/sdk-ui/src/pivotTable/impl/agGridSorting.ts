// (C) 2007-2019 GoodData Corporation
import { getIdsFromUri, getParsedFields } from "./agGridUtils";
import { FIELD_SEPARATOR, FIELD_TYPE_ATTRIBUTE, FIELD_TYPE_MEASURE, ID_SEPARATOR } from "./agGridConst";
import { assortDimensionHeaders, identifyResponseHeader } from "./agGridHeaders";
import { ISortModelItem } from "./agGridTypes";
import { ColDef } from "ag-grid-community";
import {
    IAttributeSortItem,
    IMeasureSortItem,
    isMeasureLocator,
    SortDirection,
    SortItem,
} from "@gooddata/sdk-model";
import { IAttributeHeader, IExecutionResult, IMeasureHeaderItem } from "@gooddata/sdk-backend-spi";
import invariant = require("invariant");

/*
 * All code related to sorting the ag-grid backed Pivot Table is concentrated here
 */

export const getAttributeSortItemFieldAndDirection = (
    sortItem: IAttributeSortItem,
    attributeHeaders: IAttributeHeader[],
): [string, string] => {
    const localIdentifier = sortItem.attributeSortItem.attributeIdentifier;

    const sortHeader = attributeHeaders.find(
        header => header.attributeHeader.localIdentifier === localIdentifier,
    );
    invariant(sortHeader, `Could not find sortHeader with localIdentifier ${localIdentifier}`);

    const field = identifyResponseHeader(sortHeader);
    return [field, sortItem.attributeSortItem.direction];
};

export const getMeasureSortItemFieldAndDirection = (
    sortItem: IMeasureSortItem,
    measureHeaderItems: IMeasureHeaderItem[],
): [string, string] => {
    const keys: string[] = [];
    sortItem.measureSortItem.locators.forEach(locator => {
        if (isMeasureLocator(locator)) {
            const measureSortHeaderIndex = measureHeaderItems.findIndex(
                measureHeaderItem =>
                    measureHeaderItem.measureHeaderItem.localIdentifier ===
                    locator.measureLocatorItem.measureIdentifier,
            );
            keys.push(`m${ID_SEPARATOR}${measureSortHeaderIndex}`);
        } else {
            const key = `a${ID_SEPARATOR}${getIdsFromUri(locator.attributeLocatorItem.element).join(
                ID_SEPARATOR,
            )}`;
            keys.push(key);
        }
    });
    const field = keys.join(FIELD_SEPARATOR);
    const direction = sortItem.measureSortItem.direction;
    return [field, direction];
};

export const getSortItemByColId = (
    result: IExecutionResult,
    colId: string,
    direction: SortDirection,
): IMeasureSortItem | IAttributeSortItem => {
    const { dimensions } = result;

    const fields = getParsedFields(colId);
    const [lastFieldType, lastFieldId] = fields[fields.length - 1];

    // search columns first when sorting in columns to use the proper header
    // in case the same attribute is in both rows and columns
    const searchDimensionIndex = lastFieldType === FIELD_TYPE_MEASURE ? 1 : 0;
    const { attributeHeaders, measureHeaderItems } = assortDimensionHeaders([
        dimensions[searchDimensionIndex],
    ]);

    if (lastFieldType === FIELD_TYPE_ATTRIBUTE) {
        for (const header of attributeHeaders) {
            if (getIdsFromUri(header.attributeHeader.uri)[0] === lastFieldId) {
                return {
                    attributeSortItem: {
                        direction,
                        attributeIdentifier: header.attributeHeader.localIdentifier,
                    },
                };
            }
        }
        invariant(false, `could not find attribute header matching ${colId}`);
    } else if (lastFieldType === FIELD_TYPE_MEASURE) {
        const headerItem = measureHeaderItems[parseInt(lastFieldId, 10)];
        const attributeLocators = fields.slice(0, -1).map((field: string[]) => {
            // first item is type which should be always 'a'
            const [, fieldId, fieldValueId] = field;
            const attributeHeaderMatch = attributeHeaders.find((attributeHeader: IAttributeHeader) => {
                return getIdsFromUri(attributeHeader.attributeHeader.formOf.uri)[0] === fieldId;
            });
            invariant(
                attributeHeaderMatch,
                `Could not find matching attribute header to field ${field.join(ID_SEPARATOR)}`,
            );
            return {
                attributeLocatorItem: {
                    attributeIdentifier: attributeHeaderMatch.attributeHeader.localIdentifier,
                    element: `${attributeHeaderMatch.attributeHeader.formOf.uri}/elements?id=${fieldValueId}`,
                },
            };
        });
        return {
            measureSortItem: {
                direction,
                locators: [
                    ...attributeLocators,
                    {
                        measureLocatorItem: {
                            measureIdentifier: headerItem.measureHeaderItem.localIdentifier,
                        },
                    },
                ],
            },
        };
    }
    invariant(false, `could not find header matching ${colId}`);
};

export function getSortsFromModel(sortModel: ISortModelItem[], result: IExecutionResult): SortItem[] {
    return sortModel.map((sortModelItem: ISortModelItem) => {
        const { colId, sort } = sortModelItem;
        const sortHeader = getSortItemByColId(result, colId, sort);
        invariant(sortHeader, `unable to find sort item by field ${colId}`);
        return sortHeader;
    });
}

export const assignSorting = (colDef: ColDef, sortingMap: { [key: string]: string }): void => {
    const direction = sortingMap[colDef.field];
    if (direction) {
        colDef.sort = direction;
    }
};
