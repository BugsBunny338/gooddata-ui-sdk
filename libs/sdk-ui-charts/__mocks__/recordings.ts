// (C) 2020 GoodData Corporation

import { DataViewFacade } from "@gooddata/sdk-backend-spi";
import { ScenarioRecording, recordedDataView } from "@gooddata/sdk-backend-mockingbird";

export function recordedDataFacade(rec: ScenarioRecording): DataViewFacade {
    return new DataViewFacade(recordedDataView(rec));
}
