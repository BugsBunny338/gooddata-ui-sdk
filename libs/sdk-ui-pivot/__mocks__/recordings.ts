// (C) 2020 GoodData Corporation

import { DataViewFacade } from "@gooddata/sdk-backend-spi";
import { ScenarioRecording, recordedDataView, DataViewAll } from "@gooddata/sdk-backend-mockingbird";

export function recordedDataFacade(rec: ScenarioRecording, dataViewId: string = DataViewAll): DataViewFacade {
    return new DataViewFacade(recordedDataView(rec, dataViewId));
}
