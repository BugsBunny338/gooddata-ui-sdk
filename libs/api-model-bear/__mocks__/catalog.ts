// (C) 2019 GoodData Corporation
import { GdcCatalog } from "../src";

export const catalogAttribute: GdcCatalog.ICatalogAttribute = {
    identifier: "attr",
    type: "attribute",
    title: "Attribute",
    summary: "",
    groups: [],
    production: true,
    links: {
        defaultDisplayForm: "/gdc/md/project/obj/1",
        self: "/gdc/md/project/obj/0",
    },
};

export const catalogMetric: GdcCatalog.ICatalogMetric = {
    identifier: "measure",
    type: "metric",
    title: "Measure",
    summary: "",
    groups: [],
    production: true,
    expression: "SELECT SUM([/gdc/md/project/obj/4])",
    format: "#,##",
    links: {
        self: "/gdc/md/project/obj/2",
    },
};

export const catalogFact: GdcCatalog.ICatalogFact = {
    identifier: "fact",
    type: "fact",
    title: "Fact",
    summary: "",
    groups: [],
    production: true,
    links: {
        self: "/gdc/md/project/obj/3",
    },
};
