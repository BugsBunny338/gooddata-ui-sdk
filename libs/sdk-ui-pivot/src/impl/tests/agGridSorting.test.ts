// (C) 2007-2020 GoodData Corporation

import { getSortItemByColId, getSortsFromModel } from "../agGridSorting";
import { DataViewFirstPage } from "@gooddata/sdk-backend-mockingbird";
import { ReferenceLdm, ReferenceRecordings } from "@gooddata/reference-workspace";
import { recordedDataFacade } from "../../../__mocks__/recordings";
import { newAttributeAreaSort } from "@gooddata/sdk-model";

describe("getSortItemByColId", () => {
    const fixture = recordedDataFacade(
        ReferenceRecordings.Scenarios.PivotTable.SingleMeasureWithRowAndColumnAttributes,
        DataViewFirstPage,
    );

    it("should return an attributeSortItem", () => {
        expect(getSortItemByColId(fixture.result(), "a_1055", "asc")).toMatchSnapshot();
    });

    it("should return an attributeSortItem with aggregation", () => {
        const originalSorts = [newAttributeAreaSort(ReferenceLdm.Product.Name, "desc")];

        expect(getSortItemByColId(fixture.result(), "a_1055", "desc", originalSorts)).toEqual(
            newAttributeAreaSort(ReferenceLdm.Product.Name, "desc"),
        );
    });
    it("should return an attributeSortItem with aggregation with different direction", () => {
        const originalSorts = [newAttributeAreaSort(ReferenceLdm.Product.Name, "asc")];

        expect(getSortItemByColId(fixture.result(), "a_1055", "desc", originalSorts)).toEqual(
            newAttributeAreaSort(ReferenceLdm.Product.Name, "desc"),
        );
    });

    it("should return a measureSortItem", () => {
        expect(getSortItemByColId(fixture.result(), "a_1086_460488-m_0", "asc")).toMatchSnapshot();
    });
});

describe("getSortsFromModel", () => {
    const fixture = recordedDataFacade(
        ReferenceRecordings.Scenarios.PivotTable.SingleMeasureWithRowAndColumnAttributes,
        DataViewFirstPage,
    );

    it("should return sortItems for row attribute sort", () => {
        const sortModel: any[] = [
            {
                colId: "a_1055",
                sort: "asc",
            },
        ];
        expect(getSortsFromModel(sortModel, fixture.result())).toMatchSnapshot();
    });
    it("should return sortItems for measure sort", () => {
        const sortModel: any[] = [
            {
                colId: "a_1086_460488-m_0",
                sort: "asc",
            },
        ];
        expect(getSortsFromModel(sortModel, fixture.result())).toMatchSnapshot();
    });
});
