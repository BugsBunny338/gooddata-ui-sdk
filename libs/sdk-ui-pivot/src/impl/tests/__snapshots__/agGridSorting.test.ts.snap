// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSortItemByColId should return a measureSortItem 1`] = `
Object {
  "measureSortItem": Object {
    "direction": "asc",
    "locators": Array [
      Object {
        "attributeLocatorItem": Object {
          "attributeIdentifier": "a_label.owner.region",
          "element": "/gdc/md/referenceworkspace/obj/1086/elements?id=460488",
        },
      },
      Object {
        "measureLocatorItem": Object {
          "measureIdentifier": "m_aangOxLSeztu",
        },
      },
    ],
  },
}
`;

exports[`getSortItemByColId should return an attributeSortItem 1`] = `
Object {
  "attributeSortItem": Object {
    "attributeIdentifier": "a_label.product.id.name",
    "direction": "asc",
  },
}
`;

exports[`getSortsFromModel should return sortItems for measure sort 1`] = `
Array [
  Object {
    "measureSortItem": Object {
      "direction": "asc",
      "locators": Array [
        Object {
          "attributeLocatorItem": Object {
            "attributeIdentifier": "a_label.owner.region",
            "element": "/gdc/md/referenceworkspace/obj/1086/elements?id=460488",
          },
        },
        Object {
          "measureLocatorItem": Object {
            "measureIdentifier": "m_aangOxLSeztu",
          },
        },
      ],
    },
  },
]
`;

exports[`getSortsFromModel should return sortItems for row attribute sort 1`] = `
Array [
  Object {
    "attributeSortItem": Object {
      "attributeIdentifier": "a_label.product.id.name",
      "direction": "asc",
    },
  },
]
`;
